Sources/AsyncHTTPClient/ConnectionPool/HTTP2/HTTP2IdleHandler.swift

```diff
@@ -0,0 +1,170 @@
+    }
+
+    func channelActive(context: ChannelHandlerContext) {
+        self.state.connected()
+        context.fireChannelActive()
+    }
+
+    func channelRead(context: ChannelHandlerContext, data: NIOAny) {
+        let frame = self.unwrapInboundIn(data)
+
+        switch frame.payload {
+        case .goAway:
+            let action = self.state.goAwayReceived()
+            self.run(action, context: context)
+        case .settings(.settings(let settings)):
+            let action = self.state.settingsReceived(settings)
+            self.run(action, context: context)
+        default:
+            // We're not interested in other events.
+            ()
+        }
+
+        context.fireChannelRead(data)
+    }
+
+    func userInboundEventTriggered(context: ChannelHandlerContext, event: Any) {
+        switch event {
+        case is NIOHTTP2StreamCreatedEvent:
+            let action = self.state.streamCreated()
+            self.run(action, context: context)
+            context.fireUserInboundEventTriggered(event)
+        case is NIOHTTP2.StreamClosedEvent:
+            let action = self.state.streamClosed()
+            self.run(action, context: context)
+            context.fireUserInboundEventTriggered(event)
+        default:
+            context.fireUserInboundEventTriggered(event)
+        }
+    }
+
+    func run(_ action: StateMachine.Action, context: ChannelHandlerContext) {
+        switch action {
+        case .nothing:
+            break
+        case .notifyConnectionNewSettings(let settings):
+            self.connection.http2SettingsReceived(settings)
+        case .notifyConnectionStreamClosed(let currentlyAvailable):
+            self.connection.http2StreamClosed(availableStreams: currentlyAvailable)
+        case .notifyConnectionGoAwayReceived:
+            self.connection.http2GoAwayReceived()
+        }
+    }
+}
+
+extension HTTP2IdleHandler {
+    struct StateMachine {
+        enum Action {
+            case notifyConnectionNewSettings(HTTP2Settings)
+            case notifyConnectionGoAwayReceived
+            case notifyConnectionStreamClosed(currentlyAvailable: Int)
+            case nothing
+        }
+
+        enum State {
+            case initialized
+            case connected
+            case active(openStreams: Int, maxStreams: Int)
+            case goAwayReceived(openStreams: Int, maxStreams: Int)
+            case closed
+        }
+
+        var state: State = .initialized
+
+        mutating func connected() {
+            guard case .initialized = self.state else {
+                preconditionFailure("Invalid state")
+            }
+
+            self.state = .connected
+        }
+
+        mutating func settingsReceived(_ settings: HTTP2Settings) -> Action {
+            guard case .connected = self.state else {
+                preconditionFailure("Invalid state")
+            }
+
+            let maxStream = settings.first(where: { $0.parameter == .maxConcurrentStreams })?.value ?? 100
```

Comment:
I don't believe it's invalid for there to be more than one entry per setting, i.e. it's possible for `.maxConcurrentStreams` to appear more than once in this array. Settings must however be processed in order so we should use `last(where:)` instead.